### PR TITLE
Json: Provide better error message when 'settings' is not an object

### LINKED ISSUE #5457: StandardJSON crashes on invalid source with Json::Value::find
Found by solfuzzer.

```
{
  "language": "Solidity",
  "sources": {
    "ol": {
  "content": "co]tract myfunction oill() { if
(msg.sender == owner) selfdestruct(owner); } }"  }, "mortal": {   "content": "contract mortal is owned { function kill() { if
(msg.sender == owner) selfdestruct(owner); } }"
    }
  },
  "settings":3333333333333333333333333}
```

### LINKED ISSUE #5458: StandardJSON crashes on invalid input for outputSelection
Found by solfuzzer.

Code is here: https://github.com/ethereum/solidity/blob/develop/libsolidity/interface/StandardCompiler.cpp#L97

It crashes if the the second layer ("contracts") is not an array.

### PR DESCRIPTION
**Pull request by Marenz**
Fixes #5457 
Fixes #5458

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ Compiler Features:
  * Inline Assembly: Improve error messages around invalid function argument count.
  * Code Generator: Use codecopy for string constants more aggressively.
  * Code Generator: Use binary search for dispatch function if more efficient. The size/speed tradeoff can be tuned using ``--optimize-runs``.
+ * Commandline interface: Report better error message for json input when 'settings' is not an object
```
Comment (chriseth) [inline Changelog.md]: I would say preventing an internal compiler error from happening should be a bugfix and not a feature.

**Comment by chriseth** [comment]
I think it makes sense to check that all the items in the json are of the right kind (object, string, array, ...) in a single PR. Could you please add the other ones?

**Comment by Marenz** [comment]
Pushed an update:
* Added tests for every new check
* Added checks for every part I could find

File: libsolidity/interface/StandardCompiler.cpp [hunk 2]
```diff
@@ -228,48 +228,93 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
 	return output;
 }
 
-boost::optional<Json::Value> checkKeys(Json::Value const& _input, set<string> const& _keys)
+boost::optional<Json::Value> checkKeys(Json::Value const& _input, set<string> const& _keys, const string& _name )
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]:
```suggestion
boost::optional<Json::Value> checkKeys(Json::Value const& _input, set<string> const& _keys, const string& _name)
```

File: libsolidity/interface/StandardCompiler.cpp [hunk 3]
```diff
@@ -228,48 +228,93 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
...
 {
+	if (!!_input && !_input.isObject())
+		return formatFatalError("JSONError", "\"" + _name+ "\" must be an object");
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]:
```suggestion
		return formatFatalError("JSONError", "\"" + _name + "\" must be an object");
```

File: libsolidity/interface/StandardCompiler.cpp [hunk 4]
```diff
@@ -228,48 +228,93 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
...
 
 boost::optional<Json::Value> checkRootKeys(Json::Value const& _input)
 {
 	static set<string> keys{"auxiliaryInput", "language", "settings", "sources"};
-	return checkKeys(_input, keys);
+	return checkKeys(_input, keys, "root");
 }
 
-boost::optional<Json::Value> checkSourceKeys(Json::Value const& _input)
+boost::optional<Json::Value> checkSourceKeys(Json::Value const& _input, const string& _name)
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]:
```suggestion
boost::optional<Json::Value> checkSourceKeys(Json::Value const& _input, string const& _name)
```

**Review by chriseth — CHANGES_REQUESTED** [review]
Fine apart from the formatting changes.

**Comment by Marenz** [comment]
Fixed mentioned issues

**Comment by Marenz** [comment]
Fixed failing build (misplaced `&` in last fix)

**Review by chriseth — APPROVED** [review]

**Pull request merged**